crates/astria-sequencer/src/accounts/state_ext.rs

```diff
@@ -69,8 +74,47 @@ fn nonce_storage_key<T: AddressBytes>(address: T) -> String {
 pub(crate) trait StateReadExt: StateRead + crate::assets::StateReadExt {
     #[instrument(skip_all)]
     async fn get_account_balances(&self, address: Address) -> Result<Vec<AssetBalance>> {
+        let ibc_prefixed_balances = self
+            .get_account_balances_ibc_prefixed(address)
+            .await
+            .context("failed to grab ibc balances for account")?;
+        let mut balances: Vec<AssetBalance> = Vec::with_capacity(ibc_prefixed_balances.len());
+
+        for (asset, balance) in ibc_prefixed_balances {
+            let native_asset = self
+                .get_native_asset()
+                .await
+                .context("failed to read native asset from state")?;
+            if asset == native_asset.to_ibc_prefixed() {
+                balances.push(AssetBalance {
+                    denom: native_asset.into(),
+                    balance,
+                });
+                continue;
+            }
+
+            let denom = self
+                .map_ibc_to_trace_prefixed_asset(asset)
+                .await
+                .context("failed to get ibc asset denom")?
+                .context("asset denom not found when user has balance of it; this is a bug")?
+                .into();
+            balances.push(AssetBalance {
+                denom,
+                balance,
+            });
+        }
+
+        Ok(balances)
+    }
+
+    #[instrument(skip_all)]
+    async fn get_account_balances_ibc_prefixed<T: AddressBytes>(
+        &self,
+        address: T,
+    ) -> Result<HashMap<IbcPrefixed, u128>> {
         let prefix = format!("{}/balance/", StorageKey(&address));
-        let mut balances: Vec<AssetBalance> = Vec::new();
+        let mut balances: HashMap<IbcPrefixed, u128> = HashMap::new();
```

Comment:
This should be an `IndexMap` to provide a stable order.

Reply:
We don't really need the balances sorted in the rest of the code that uses the `get_account_balances_ibc_prefixed()` function. I'd prefer to just sort the returned vec in either `get_account_balance()` or in the one function that calls it (`balance_request()`)
https://github.com/astriaorg/astria/blob/e9d5f0f0fc2f2449292aeeb07a1f2f7e7227a637/crates/astria-sequencer/src/accounts/query.rs#L36

Reply:
The point is that you get the sorting for free (in the same order as the Vec).

Since indexmap is already a dependency this is doesn't have downsides.